File: api/lib/spotify.rb
# frozen_string_literal: true

class Spotify
  def initialize
    RSpotify.authenticate(ENV['SPOTIFY_CLIENT_ID'], ENV['SPOTIFY_CLIENT_SECRET'])
  end

  def top_tracks_by_year(year)
    RSpotify::Track.search("year:#{year}").map { |track| format_track(track) }
  end

  private

  def format_track(track)
    {
      id: track.id,
      artist: track.artists.map(&:name).join(', '),
      title: track.name,
      image: track.album.images.max_by { |image| image['height'] }['url'],
      preview: track.preview_url
    }
  end
end

File: api/src/services/providers/spotify.ts
import SpotifyWebApi = require("spotify-web-api-node");

import { ITracksProvider } from "../../contracts/ITracksProvider";
import { ITrack } from "../../models/ITrack";

const spotifyWebApi = new SpotifyWebApi({
  clientId: process.env.SPOTIFY_CLIENT_ID,
  clientSecret: process.env.SPOTIFY_CLIENT_SECRET
});

export class Spotify implements ITracksProvider {
  async getTracksForYear(year: number): Promise<ITrack[]> {
    const credentials = await spotifyWebApi.clientCredentialsGrant();
    spotifyWebApi.setAccessToken(credentials.body.access_token);

    const response = await spotifyWebApi.searchTracks(`year:${year}`);

    return response.body.tracks.items.map(x => ({
      artist: x.artists.map(y => y.name).join(' '),
      title: x.name,
      id: x.id,
      image: x.album.images[0].url,
      preview: x.preview_url
    }) as ITrack);
  }
}